### PR TITLE
fix(areal): saturate past maxpos instead of wrapping into the sign bit

### LINKED ISSUE #1503: areal: values just above the largest finite exponent come out with the wrong sign or as NaN instead of saturating
## Summary

areal arithmetic and conversion give wrong-sign or NaN results for any value in the binade just above the largest finite exponent, and for values in the top binade above `maxpos`. They should saturate to `(maxpos, inf)` / `(maxneg, -inf)` with the ubit set, as larger overflows already do.

Measured on `main` (5a0ac4cb), gcc 13.3:

| expression | result | expected |
|---|---|---|
| `areal<16,5>`: `maxpos + maxpos` (261632) | `[-6.07967e-05]` | `(130816, inf)` |
| `areal<16,5>`: `65536 * 2` (131072) | `[-0]` | `(130816, inf)` |
| `areal<16,5>`: `65536 + 65472` (131008) | `[nan]` | `(130816, inf)` |
| `areal<16,5>`: `maxpos * 1.001953125` | `[nan]` | `(130816, inf)` |
| `areal<16,5>(131000.0)` | `[nan]` | `(130816, inf)` |
| `areal<16,5>(131072.0)` | `[-0]` | `(130816, inf)` |
| `areal<32,8>`: `maxpos * 2` | `[-1.17549e-38]` | `(maxpos, inf)` |

Larger values are fine: `areal<16,5>(1e10)` gives `(130816, inf)`.

## Two causes, in both conversion paths

The same logic appears in `operator=(double)` and in `convert(blocktriple, areal, inputUncertain)` in `areal_impl.hpp`. The second is what every arithmetic operator ends in.

1. **`MAX_EXP` is one too large.** `MAX_EXP = (1 << es) - EXP_BIAS`, which is 17 for `es = 5`, `bias = 15`. The largest exponent field, 31, gives scale 31 - 15 = 16. A value of scale 17 passes the `exponent > MAX_EXP` saturation test and gets the biased exponent 32. That overflows the 5-bit field into the sign bit, giving `-0`, or `-6.08e-05` once the fraction bits land in the exponent. Scales 18 and up saturate correctly.
2. **The top binade does not stop at `maxpos`.** At scale 16, `maxpos` has fraction `111111110`. The all-ones fraction is reserved: `inf` with the ubit clear, `nan` with it set. A value from `maxpos + ulp` up to 2^17 truncates onto those patterns and becomes `inf` or `nan`, where it should saturate.

## Suggested fix

In both paths, saturate to `(maxpos, inf)` / `(maxneg, -inf)` with the ubit set when the scale exceeds the largest finite scale, or when it equals it and the truncated fraction lies above `maxpos`'s. A regression test should drive every operator and conversion across the top two binades against `double`, for a few configurations including a multi-limb one.

Found while implementing `operator>>` for #1454. The new parser goes through the same `convert()`, so decimal inputs in this band inherit the bug until it is fixed.

### PR DESCRIPTION
**Pull request by Ravenwater**
## Summary

Values just above the largest finite exponent came out with the wrong sign or as nan instead of saturating to `(maxpos, inf)` (#1503). On `areal<16,5>`, `maxpos + maxpos` gave `-6.08e-05`, `65536 * 2` gave `-0`, and `65536 + 65472` gave `nan`. There were two causes, and both affected every conversion:

1. **`MAX_EXP` was one too large.** It was `2^es - EXP_BIAS` when the largest finite scale is `2^es - 1 - EXP_BIAS`. A value at that scale passed the overflow checks, and its biased exponent ran into the sign bit.
2. **Truncation in the top binade hits reserved encodings.** Every value past `maxpos + ulp` truncates onto the all-ones fraction, which encodes inf (ubit clear) or nan (ubit set).
   - The new `saturate_reserved()` maps a finite result that lands there to `(maxpos, inf)` or `(maxneg, -inf)`, the interval that larger values already get.
   - It runs at the end of the integer, `float` and `double` assignments and of `convert(blocktriple)`, so `+ - * /` and `parse()` saturate too.

### Conversion from `long double` (CodeRabbit review)

`operator=(long double)` narrowed to `double` first, which rounds:
- a finite value past `DBL_MAX` became inf;
- a value below double's range became an exact zero;
- bits below double's precision were gone before truncation could see them, so `1 + 2^-60` and `1 - 2^-60` both read as exactly `[1]`.

Where long double is wider than double, the new `assign_extended()`:
- normalizes `|x|` into [1, 2) by powers of two, which is exact and stays `constexpr` (no `frexp`);
- packs the top 64 significand bits into a `blocktriple` the way `parse()` does;
- folds any lower bits into the ubit, and lets `convert()` truncate and saturate.

The path is selected on `numeric_limits<long double>::digits`, not `LONG_DOUBLE_SUPPORT`. That macro is 1 on Apple arm64, where long double is double, and 0 on RISC-V, where it is IEEE quad.

### NaN payloads (CodeRabbit review)

`operator=(float)` and `operator=(double)` recognized only the two canonical NaN encodings. Every other payload fell through to the numeric path and came out as a number, already on main:
- `(maxpos, inf)` on `areal<16,5>`;
- `[inf]` on `areal<64,11>`;
- a finite `5.1e38` for a float NaN on `areal<32,8>`.

Now an all-ones exponent is inf with a zero fraction and a NaN with any other, and the quiet bit sets the kind. This is the #1303 rule `blocktriple` already follows.

## Test

`static/range/areal/conversion/saturation.cpp` checks results against a reference that uses no areal conversion. The reference enumerates every encoding of a small configuration and tabulates its exact value. A finite `x` then maps to one of:
- the exact value equal to it;
- the exact value below it in magnitude, with the ubit set;
- `(maxpos, inf)` past the top;
- `(0, minpos)` below minpos.

Each result carries the sign of `x`.

| group | configurations | on 1700f2c1 (main) | this branch |
|---|---|---|---|
| the #1503 cases, `parse("131000")`, `areal<32,8>`, `areal<64,6>` native integer path, `areal<128,15>` | | 14 fail | pass |
| conversion from double and float: every exact value and midpoint, the top binade and beyond, both signs | `<8,2>`, `<10,3,u8>`, `<12,4,u16>`, `<16,5,u16>`, `<16,5,u8>` | 288 fail each | pass |
| `+ - *` over every pair of exact values | `<8,2>`, `<9,3,u16>`, `<10,3,u16>` (L1), `<12,5>` (L2), `<12,4>` (L3), `<13,4>` (L4) | 5892 / 8292 / 34084 at L1 | pass |
| conversion from long double: every exact value, and `2^-58` of its scale above and below | `<8,2>`, `<12,4,u16>`, `<16,5,u16>`, `<16,5,u8>` | 252 / 4092 / 65532 / 65532 fail | pass |
| NaN payloads, quiet and signalling, both signs, from float and double | `<16,5,u16>`, `<16,5,u8>`, `<32,8>`, `<64,11>` | 24 of 32 fail each | pass |
| long double: `areal<128,15>` encodings of 64-bit significands and `2^+-2000` (frexp oracle); `areal<64,11>(1 +/- 2^-60)`; constexpr conversion | | 8 fail | pass |

- **Runtime:** level 1 takes 0.15 s; all four levels take about 10 s.
- **Verified:**
  - All levels pass on gcc 13 and clang.
  - The whole test also passes under QEMU on aarch64 and riscv64, where long double is IEEE quad, and the long double conversion constant-evaluates on all four targets.
  - The areal suite passes 25/25 on both compilers, with no warnings.
  - The other areal consumers (the `applications/precision/ubit` programs, `numeric`, `static/utility`, `education`, `tools/cmd/areal`) print identical output on this branch and on main.

## Found along the way, filed separately

- **#1506:** `+ - *` are wrong across the range when the exponent field straddles two limbs (for example `areal<10,3,uint8_t>`, `areal<32,8,uint8_t>`). The arithmetic check here uses single-limb configurations until that is fixed. Multi-limb results are the same on main and on this branch.
- **#1507:** `+0 - +0` gives `-0`. The arithmetic check accepts either sign on an exact zero.
- **#1509:** conversion *to* `long double` gives `-nan` on clang and loses range beyond double's on gcc. The long double checks here test encodings rather than round-tripping.

## Behaviour change

- Conversions and arithmetic that overflow the top binade now give `(maxpos, inf)` or `(maxneg, -inf)` instead of a wrong-signed value, `-0`, nan or inf.
- A float or double NaN with a non-canonical payload now converts to an areal NaN of the same kind instead of `(maxpos, inf)`, inf, or a finite value.
- Assigning a `long double` no longer rounds through `double`: values beyond double's range saturate or underflow correctly, and bits below double's precision set the ubit.

Closes #1503

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01NazYbxaii2XXVzQeQKMDTs

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved upper-range handling for areal number conversions and arithmetic.
  - Values exceeding supported ranges now saturate to the appropriate signed maximum with the uncertainty indicator set.
  - Applied consistently across integer, floating-point, and general conversions.
  - Improved quiet and signalling NaN classification for float and double conversions.
  - Preserved extended precision for supported `long double` values, including values beyond double’s range.

- **Tests**
  - Added comprehensive regression coverage for boundary values, overflow, conversions, arithmetic, NaN payloads, and wider-precision calculations.

### CODE REVIEW

**Comment by Ravenwater** [comment]
@coderabbitai Thanks for the long double finding (areal_impl.hpp `operator=(long double)`, outside the diff). It was valid, and it went further than the report: narrowing through `double` also turned a value below double's range into an exact zero, and dropped bits below double's precision before truncation could see them. So `1 + 2^-60` and `1 - 2^-60` both read as exactly `[1]` instead of `(1, next)` and `(prev, 1)`.

Fixed in f2e58beb, with two departures from the suggestion.

**No `blocktriple` long-double conversion.** `blocktriple::convert_ieee754` rounds the significand, where areal needs truncation plus a ubit, and it leaves subnormals as TBD. The new `assign_extended()` instead:
- normalizes `|x|` into [1, 2) by powers of two, which is exact and needs no `frexp`, so the conversion stays `constexpr` (`api/constexpr.cpp` constructs `areal(1.0l)` in a constant expression);
- packs the top 64 significand bits into a `blocktriple<fbits, MUL>` the way `parse()` does;
- folds any bits below those into the ubit;
- then lets `convert()` truncate and saturate as it does for every other source.

**Keyed on the format, not `LONG_DOUBLE_SUPPORT`.** The path is selected on `numeric_limits<long double>::digits > numeric_limits<double>::digits`. `long_double.hpp` sets `LONG_DOUBLE_SUPPORT` to 1 on Apple arm64, where long double is double, and to 0 on RISC-V, where it is IEEE quad.

**Regression cases in `static/range/areal/conversion/saturation.cpp`:**
- every exact value, and a value `2^-58` of its scale above and below it, against the enumerated reference;
- `2^2000` and `2^-2000`;
- `areal<128,15>` encodings of 64-bit significands, checked against a `frexp`-built oracle;
- the constexpr conversion.

They pass on gcc and clang (x87) and under QEMU on aarch64 and riscv64 (quad). Against main, the long double groups fail by the thousands.

The test checks encodings rather than round-tripping, because the reverse conversion `to_native<long double>()` has its own defects: `-nan` on clang, and range capped at double's on gcc. That is filed as #1509.

**Comment by Ravenwater** [comment]
@coderabbitai Thanks for the NaN finding (`operator=(float)`/`operator=(double)` special cases, outside the diff). It was valid, and the defect was already on main, independent of `saturate_reserved()`: with only two payloads recognized, every other NaN fell into the numeric path and came out as a number there too. On 1700f2c1:
- `areal<16,5>(double 0x7FF8000000000002)` gives `(130816, inf)`;
- `areal<64,11>` of the same NaN gives `[inf]`;
- `areal<32,8>(float 0x7FC00002)` gives a finite `[5.10424e+38]`.

This PR produced the same results, since the early overflow check or the assembly caught those NaNs before `saturate_reserved()` ran.

Fixed in c363767b at the root: exponent all ones is infinity with a zero fraction, and a NaN with any other fraction, whatever the payload. The quiet bit selects `NAN_TYPE_QUIET` or `NAN_TYPE_SIGNALLING`, the same rule #1303 applied to `blocktriple`. The canonical encodings convert exactly as before.

`static/range/areal/conversion/saturation.cpp` now converts quiet and signalling payloads of both signs, from float and double, on `areal<16,5,uint16_t>`, `<16,5,uint8_t>`, `<32,8>` and `<64,11>`. It checks both NaN-ness and kind:
- this branch gets all of them right on gcc and clang (x87), and under QEMU on aarch64 and riscv64;
- main gets 24 of 32 wrong per configuration.

File: static/range/areal/conversion/saturation.cpp [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+}
+
+template<typename A>
+int VerifyNaNPayloads(bool reportTestCases) {
+	int fails = 0;
+	auto verify = [&](const std::string& what, const A& a, bool quiet) {
+		if (a.isnan(quiet ? NAN_TYPE_QUIET : NAN_TYPE_SIGNALLING)) return;
+		++fails;
+		if (reportTestCases || fails < 6)
+			std::cerr << "FAIL: " << what << " gave " << to_binary(a) << ' ' << a << ", expected a " << (quiet ? "quiet" : "signalling") << " nan\n";
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply static/range/areal/conversion/saturation.cpp]: Fixed in 56e1b3d1: the file is new in this PR, so it is formatted whole with the project .clang-format (clang-format 18). No line exceeds 120 columns at TabWidth 4. All levels still pass on gcc and clang and under QEMU on aarch64 and riscv64.

**Comment by Ravenwater** [comment]
@coderabbitai Thanks. The dispatch finding (`operator=(long double)`, outside the diff) is fixed in the commit above: the exact path is now taken when `long double` has more digits, a larger `max_exponent`, or a smaller `min_exponent` than `double`, since narrowing loses range as well as precision.

No current ABI has a long double with double's precision and a wider range: x87 extended, IEEE quad and IBM double-double all have more digits, and every other target's long double is double itself. So the chosen path is unchanged on every target. The condition now states what narrowing actually loses. The tests pass as before on gcc and clang (x87) and under QEMU on aarch64 and riscv64 (quad), and the conversion still constant-evaluates.
